cilium/CFP-30572-non-default-deny-policies.md

```diff
@@ -0,0 +1,175 @@
+**SIG: SIG-Policy**
+
+**Begin Design Discussion:** 2024-02-02
+
+**Cilium Release:** 1.16
+
+**Authors:** Casey Callendrello <cdc@isovalent.com>
+
+## Summary
+
+Allow users to create network policies that do not drop non-matching traffic. 
+
+## Motivation
+
+Consider a large, multitenant cluster. The cluster administrator is distinct from
+the workload owners, and the workloads are diverse and existing.
+
+The administrator may wish to create a Cilium Clusterwide Network Policy (CCNP)
+that affects most or all existing workloads. For example, they may wish to
+- Globally grant access to a security scanner
+- Ensure that workloads can access kube-dns
+- Proxy all DNS requests for monitoring purposes
+- Deny access to sensitive IPs, such as cloud provider metadata APIs
+
+All of these can already be accomplished by the existing policy language.
+
+By default, all Cilium endpoints have unrestricted network access. If at least one
+network policy applies to an endpoint, then all traffic is denied except that
+explicitly allowed by policy.
+
+This presents a challenge for the cluster administrator. They cannot assume that
+existing, heterogeneous worloads already have network policies provisioned. So, by
+creating an administrative network policy, they may select endpoints that **previously
+were in an allow-all state**. Thus, they may unintentionally block legitimate
+traffic for an existing workload, causing an outage.
+
+## Goals
+
+* Ability to safely roll-out broad-based network policies, without the risk
+  of disrupting existing traffic.
+
+## Non-Goals
+
+* An iterative, iptables-style firewall rules engine
+* Arbitrary, multi-level rule priority
+
+## Proposal
+
+### Overview
+
+We propose a new field in the policy spec, `DefaultAction`. This field specifies
+whether or not an implicit default-deny rule should be created _on behalf of this policy_.
+
+This field is optional. The default value is `Deny`. This matches the existing behavior, wherein any network policy implicitly blocks all other traffic.
+
+The value `Pass` indicates that this policy should not create a default-deny rule.
+
+Note that this is unrelated to the `IngressDeny` and `EgressDeny` policy expressions. Those dictate the action for *matching* traffic, whereas the `DefaultAction` specifies what should happen with *non-matching* traffic.
+
+### Per-direction state
+
+Endpoints have a default rule *for each direction*. That is to say, if an endpoint only has rules that select ingress traffic, then ingress is default-deny, and egress is default-allow.
+
+The rest of this document, for shorthand, assumes all rules operate in the same direction. The logic described is identical but independent for each direction.
+
+### Precedence
+
+When an endpoint is selected by multiple rules, `Deny` shall take precedence over `Pass`. That is to say, if a new `Pass`-type rule selects an endpoint with existing policies, no new traffic shall be allowed except that selected by the new rule.
```

Comment:
Oh, good point. We use "Rule" and "Policy" somewhat interchangeably in the codebase, since a `CiliumNetworkPolicy.Spec` is just a `struct Rule`. And then `Rule.Ingress []IngressRule` further complicates matters.

This should be a policy-level field. I'll update the doc.